[FLINK-33450][autoscaler] Support the JDBCAutoScalerStateStore

1996fanrui opened the pull request:
## What is the purpose of the change

Introduced the JDBCAutoScalerStateStore


## Brief change log

- [FLINK-33450][autoscaler] Extract the AbstractAutoScalerStateStoreTest and add the InMemoryAutoScalerStateStoreTest
  - JDBCAutoScalerStateStore can reuse the AbstractAutoScalerStateStoreTest as well 
- [FLINK-33450][autoscaler] Support the JDBCAutoScalerStateStore
  - Introduce a series of `DatabaseTest` in this commit as well, it's easy to test and finish the ITCase for derby, mysql and PostgreSQL
  - JDBCAutoScalerStateStore has cache as well, it's similar to KubernetesAutoScalerStateStore, core class:
  - StateType, JDBCStateInteractor, JDBCStore and JobStateView
- [FLINK-33450][autoscaler] StandaloneAutoscalerEntrypoint supports the JDBCAutoScalerStateStore
  - Introduced some options to load JDBCAutoScalerStateStore, such as:
  - autoscaler.standalone.state-store.type
  - autoscaler.standalone.state-store.jdbc.password
  - autoscaler.standalone.state-store.jdbc.url
  - autoscaler.standalone.state-store.jdbc.username
- [FLINK-33450][autoscaler] Document the JDBC Autoscaler State Store

## Verifying this change

- Added a series of unit tests and ITCases in `flink-autoscaler-plugin-jdbc` module
- Autoscaler Standalone uses mysql and PostgreSQL manually, it works well

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes, added some dependencies to `flink-autoscaler-plugin-jdbc` module
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: no

## Documentation

  - Generated the configuration option docs
  - Improve the autoscaler standalone doc

1996fanrui reviewed (commented):
Hi @gyfora @mxm , the `JDBCAutoScalerStateStore` is supported in this PR, and I have tested it manually with Autoscaler standalone, it works well.

Please take a look in your free time, thanks a lot~ ❤️

mxm commented on flink-autoscaler-plugin-jdbc/src/main/java/org/apache/flink/autoscaler/jdbc/state/JobStateView.java:
```diff
@@ -0,0 +1,262 @@
+
+/** The view of job state. */
+@NotThreadSafe
+public class JobStateView {
+
+    /**
+     * The state of state type about the cache and database.
+     *
+     * <p>Note: {@link #inLocally} and {@link #inDatabase} are only for understand, we don't use
+     * them.
+     */
+    @SuppressWarnings("unused")
+    enum State {
+
+        /** State doesn't exist at database, and it's not used so far, so it's not needed. */
+        NOT_NEEDED(false, false, false),
+        /** State is only stored locally, not created in JDBC database yet. */
+        NEEDS_CREATE(true, false, true),
+        /** State exists in JDBC database but there are newer local changes. */
+        NEEDS_UPDATE(true, true, true),
+        /** State is stored locally and in database, and they are same. */
+        UP_TO_DATE(true, true, false),
+        /** State is stored in database, but it's deleted in local. */
+        NEEDS_DELETE(false, true, true);
```
This looks conceptually very similar to `ConfigMapView`. I wonder if we could have shared some of that logic.

1996fanrui replied:
Yeah, I refer to `ConfigMapView` during developing. I didn't share logic due to some reasons:

1. The View and State are different
  - `ConfigMapView` stores `ConfigMap` for each job, so we create it directly.
  - `JobStateView` stores state for each stateType, it means that part of state types of current job are stored in the database, but the rest of the state types may have been created in the database.
  - That's why `JobStateView` have `NOT_NEEDED` and `NEEDS_DELETE`, but `ConfigMapView` doesn't.
2. Based on the first reason, the `ConfigMapView` accesses the whole ConfigMap, `JobStateView`  accesses each row of database. It's hard to share code.

mxm replied:
Got it. The code complexity for adding a new state backend makes we wonder though, if we need to rethink some of the building blocks for state stores. It would be good to move more of the logic into the core, to avoid duplicating / writing similar logic. I was initially more in favor of moving this out of the core but we need to see if that might be too complicated in the long run. This is out of scope for this PR though.

1996fanrui replied:
> It would be good to move more of the logic into the core, to avoid duplicating / writing similar logic. 

Sounds make sense, after I analyze, some of logic are same, such as: 

- `ConfigMapStore` and `JDBCStore` can be abstracted to `StringStateStore` interface
  - They support `put`, `get` and `remove`
  - The parameters of `ConfigMapStore` are the (JobContext, String key, String value).
  - The parameters of `JDBCStore` are the (String jobKey, StateType stateType, String value).
  - We can define a interface `StringStateStore`, and the parameters are `(JobContext, StateType stateType, String value)`.

- `KubernetesAutoScalerStateStore` and `JDBCAutoScalerStateStore` can be abstracted to `AbstractAutoscalerStateStore`
  - They support serialize and compress `Original State` to String.
  - `AbstractAutoscalerStateStore` can reuse the serialize and compress logic
  - `KubernetesAutoScalerStateStore` support the limitation of stateValue
  - We can define a parameter for `AbstractAutoscalerStateStore`, the limitation is disabled by default, and `KubernetesAutoScalerStateStore` can enable it.

And I created FLINK-34065 to follow it.

1996fanrui replied:
Hi @mxm @gyfora , in the beginning of design the state store, we use the state type id in the database table.

After I thinking, more and this suggestion from @mxm . It's better to share some code between `KubernetesAutoScalerStateStore` and `JDBCAutoScalerStateStore`. So I change the state type from id to `identifier` (From int to String).

Align it to `KubernetesAutoScalerStateStore`, using: `scalingHistory`, `scalingTracking`, `collectedMetrics` and `parallelismOverrides`. 

It has 2 advantages : 

- It's easy to share the `StateType` class
- When we use the `JDBCAutoScalerStateStore`, the identifier is easier to troubleshooting than id. For example, we needs to query  some states  from database manually. It's easy to understand.

We can ensure the `identifier` length isn't too long, it won't introduce too many storage cost.

I have update it in the latest hotfix commit, WDYT?

mxm reviewed (approved):
LGTM!

1996fanrui commented on flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/config/AutoscalerStandaloneOptions.java:
```diff
@@ -59,4 +64,47 @@ private static ConfigOptions.OptionBuilder autoscalerStandaloneConfig(String key
+            autoscalerStandaloneConfig("state-store.type")
+                    .enumType(StateStoreType.class)
+                    .defaultValue(StateStoreType.MEMORY)
+                    .withDescription("The autoscaler state store type.");
```
Emmm, I didn't list due to flink-kubernetes-operator doc will list automaticlly.

You can take a look the generated html, it has all enums of `StateStoreType`. 

And all old options have `Possible values:` as well.

<img width="996" alt="图片" src="https://github.com/apache/flink-kubernetes-operator/assets/38427477/ba2fa8b5-dc0e-4bb1-8f8f-d493c61a9bab">

mxm replied:
Got it! Forgot about that.

gyfora commented on flink-autoscaler-plugin-jdbc/src/main/java/org/apache/flink/autoscaler/jdbc/state/JDBCStore.java:
```diff
@@ -0,0 +1,92 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+/** The jdbc store. */
+public class JDBCStore {
```
Should we call it `JdbcStateStore` instead?

gyfora commented on flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/config/AutoscalerStandaloneOptions.java:
```diff
@@ -59,4 +64,47 @@ private static ConfigOptions.OptionBuilder autoscalerStandaloneConfig(String key
+                                    .build());
+
+    public static final ConfigOption<String> STATE_STORE_JDBC_PASSWORD =
+            autoscalerStandaloneConfig("state-store.jdbc.password")
```
We need a way to set this from a secret / env. Otherwise this is very hard to be used in production properly

gyfora commented on flink-autoscaler-plugin-jdbc/src/main/java/org/apache/flink/autoscaler/jdbc/state/JDBCAutoScalerStateStore.java:
```diff
@@ -0,0 +1,286 @@
+ * @param <Context> The job autoscaler context.
+ */
+@Experimental
+public class JDBCAutoScalerStateStore<KEY, Context extends JobAutoScalerContext<KEY>>
```
Should be ` JdbcAutoScalerStateStore<`

gyfora commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/state/AbstractJDBCStateInteractorITCase.java:
```diff
@@ -0,0 +1,93 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** The abstract IT case for {@link JDBCStateInteractor}. */
+abstract class AbstractJDBCStateInteractorITCase implements DatabaseTest {
```
`JDBC` -> `Jdbc
`

gyfora commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/state/AbstractJDBCStoreITCase.java:
```diff
@@ -0,0 +1,303 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+/** The abstract IT case for {@link JDBCStore}. */
+abstract class AbstractJDBCStoreITCase implements DatabaseTest {
```
JDBC -> Jdbc

gyfora commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/state/CountableJDBCStateInteractor.java:
```diff
@@ -0,0 +1,81 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Countable {@link JDBCStateInteractor}. */
+public class CountableJDBCStateInteractor extends JDBCStateInteractor {
```
JDBC -> Jdbc

gyfora commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/state/JDBCAutoScalerStateStoreTest.java:
```diff
@@ -0,0 +1,161 @@
+ * <p>Note: {@link #testCompressionMigration} and {@link #testDiscardInvalidHistory} are referred
+ * from {@link KubernetesAutoScalerStateStoreTest}.
+ */
+class JDBCAutoScalerStateStoreTest
```
JDBC -> Jdbc

gyfora reviewed (changes requested):
I added some minor comments @1996fanrui mostly for proper camel-casing the class names.

But more importantly we should allow a more flexible config of the password.

1996fanrui reviewed (commented):
Thanks @gyfora for the suggestion!

I have updated all `JDBC` to `Jdbc`. Do you have any specific solution about the password setting?

gyfora commented:
@1996fanrui would be great if Flink had environment variable resolution in the conf but an alternative would be an extra config called: `state-store.jdbc.password-env` where we could specify the env variable name to resolve. If password-env is defined then the password would be ignored or vice versa 

In Kubernetes world secrets can be easily mounted as env variables so this would make it nice and easy to configure

1996fanrui commented:
> @1996fanrui would be great if Flink had environment variable resolution in the conf but an alternative would be an extra config called: `state-store.jdbc.password-env` where we could specify the env variable name to resolve. If password-env is defined then the password would be ignored or vice versa
> 
> In Kubernetes world secrets can be easily mounted as env variables so this would make it nice and easy to configure

Thanks @gyfora for the quick feedback!

I have updated the `state-store.jdbc.password` to `state-store.jdbc.password-env-variable`, and updated the doc to guide user to export the `STATE_STORE_JDBC_PWD` env variable.

gyfora commented on docs/content/docs/custom-resource/autoscaler.md:
```diff
@@ -260,17 +260,46 @@ job.autoscaler.metrics.window : 3m
+--autoscaler.standalone.fetcher.flink-cluster.port 8081 \
+--autoscaler.standalone.state-store.type jdbc \
+--autoscaler.standalone.state-store.jdbc.url jdbc:mysql://localhost:3306/flink_autoscaler \
+--autoscaler.standalone.state-store.jdbc.username root
```
should we also state the password-env-variable config here?

1996fanrui replied:
In general, the environment variable name doesn't need to be changed. Users need to export the password using this environment variable.

So I didn't mention `password-env-variable` here. But in the beginning of this doc, it mentioned how to export password. WDYT?

```
# export the password of jdbc state store
export STATE_STORE_JDBC_PWD=123456
```

gyfora replied:
You are right, sorry I got confused by the logic for second :) 

I think it would make sense to keep the config for the password there as well as an option for people to use instead of the env. The env variable is great for prod but for testing the password may be more convenient. Wdyt?

1996fanrui replied:
Do you mean we provide the `autoscaler.standalone.state-store.jdbc.password-env-variable` and `autoscaler.standalone.state-store.jdbc.password` together?

- For prod, we recommend users to `export STATE_STORE_JDBC_PWD=xxx`.
- For testing, using `autoscaler.standalone.state-store.jdbc.password` is convenient.

Right?

I'm afraid 2 options let users get confused here. 

Also, if we provide 2 options, we need to consider the priority. Which one will be used if users set 2 options together.

1996fanrui replied:
About testing, if we only test `JdbcAutoScalerStateStore`, the password is easy to be set.

JdbcAutoScalerStateStore will work well regardless of whether the Autoscaler standalone uses  or the Kubernetes-operator uses if `JdbcAutoScalerStateStore` is tested properly. 

So test `JdbcAutoScalerStateStore` in Autoscaler standalone isn't necessary, right?

If so, we don't need the `autoscaler.standalone.state-store.jdbc.password` for now.  And we can add it if we really needs it in the future.

WDYT?

gyfora replied:
okay, sounds good

gyfora commented on flink-autoscaler-plugin-jdbc/src/main/java/org/apache/flink/autoscaler/jdbc/state/JdbcAutoScalerStateStore.java:
```diff
@@ -0,0 +1,286 @@
+        return ConfigurationUtils.convertValue(overrides, Map.class);
+    }
+
+    private static String compress(String original) throws IOException {
```
Is there really a reason to compress the data in a DB? For configmaps this was done due to the very constrained size. But compressing makes it much harder to read for users

1996fanrui replied:
No any strong reason, it refers the `KubernetesAutoScalerStateStore`.

I removed the the compression related logic for JdbcAutoScalerStateStore in the latest hotfix commit.

1996fanrui commented:
Hi @gyfora @mxm , do you have any other comments?

mxm commented:
LGTM

1996fanrui commented:
Thanks for the patient review, merging~
